doc/compiler/jitserver/Problem.md

```diff
@@ -19,9 +19,40 @@ OpenJDK Assembly Exception [2].
 
 SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 -->
+# Debugging JITServer
 
-Usually typical OpenJ9 debugging procedures can be followed for JITServer. But there are a few JITServer specific tips and tricks.
+Typical OpenJ9 debugging procedures can be followed for JITServer. But there are a few JITServer specific tips and tricks.
+Unlike regular OpenJ9, where everything runs in one process, JITServer separates JIT compiler into a separate process while constantly
+working with addresses/objects that are only valid on the client. This makes it very easy to cause a bug on JITServer and may make it hard to catch.
+In this document we will explore some common types of bugs and crashes that are encountered in JITServer development and how to deal with them.
+This document should be expanded upon whenever possible.
 
+## Server-side crashes
+The most common type of a server crash is a segmentation fault caused by the compiler trying to dereference an address that is only valid on the client.
+For example, dereferencing `J9Class` pointer will lead to immediate crash in the best case, or some corrupt data being written in the worst case.
+If you suddenly discover a crash on the server that you haven't seen before, it's most likely caused by a change to mainline JIT code that did not account
+for JITServer. If that's the case, the crash should be in the newly added code. You should find the commit that introduced it, and rewrite it in such a way
+that accounts for JITServer, i.e. do not dereference client pointers directly but fetch the result from the client first.
+
+## Client-side crashes
 If the client crashes in `handleServerMessage` after receiving some bad data from the server, you will need to find out what the server was doing when it sent the message. The easiest way to do this is by running both the server and client in gdb, then sending the server a Ctrl-C after the client crashes. From the gdb prompt on the server you can type `thread apply all backtrace` and find the appropriate compilation thread to determine where the server was.
 
-Please add to this file!!
+## Crashes in compiled code
+Another way a client can crash is when it runs the compiled method but something is wrong with it. These can be some of the tougher bugs to fix
+because you will need to find the failing method and look at its trace log to find the problem. Sometimes, the failure is not caused by one compilation
+but by a series of them, which makes debugging even harder.
+
+Tracing method compilation is the same as for non-JITServer - use `-Xjit:{<method_name>}(traceFull,log=<log_name>)` in the client options and both client
```

Comment:
Can we add somewhere in here about the double compilation debugging feature? This is the one where a method is compiled both at the server and at the client for the purpose of comparing the two compilation logs.

Reply:
Right, I actually forgot that this option exists.